rust/src/res.rs

```diff
@@ -0,0 +1,195 @@
+        if path.is_dir() {
+            for entry in path.read_dir()? {
+                if let Ok(entry) = entry {
+                    self.add_sub_path(base, &entry.path())?;
+                }
+            }
+        } else if path.is_file() {
+            let mut resource = Resource::default();
+            resource.path = my_path;
+            // include slf contents
+            if self.properties.get("with_archive_slf") == Some(&json!(true)) {
+                if uppercase_extension(&path)? == "SLF" {
+                    resource.properties.insert("archive_slf".to_string(), json!(true));
+                    let num_resources = self.add_slf(&path, &resource.path)?;
+                    resource.properties.insert("archive_slf_num_resources".to_string(), json!(num_resources));
+                }
+            }
+            self.resources.push(resource);
+        } else {
+            return Err(Box::new(ResourceError::new(format!("{:?} in not an accessible directory or file", path))));
+        }
+        return Ok(())
+    }
+
+
+    // Adds resources to the pack.
+    //
+    // The Ok entries of the SLF archive will be added as resources.
+    // The resources will have the archive_path property set.
+    // Returns the number of resources added.
+    #[allow(dead_code)]
+    fn add_slf(&mut self, path: &Path, archive_path: &str) -> Result<i32,Box<Error>>
+    {
+        let mut f = File::open(path)?;
+        let header = SlfHeader::from_input(&mut f)?;
+        f.seek(start_of_entries(header.number_of_entries))?;
+        let mut num_resources = 0;
+        for entry in header.read_entries_from(&mut f)? {
+            match entry.state {
+                SlfEntryState::Ok => {
+                    let mut resource = Resource::default();
+                    let path = header.library_path.clone() + &entry.file_name;
+                    resource.path = path.replace("\\", "/");
+                    resource.properties.insert("archive_path".to_string(), json!(archive_path));
+                    // TODO include archive inside archive?
+                    self.resources.push(resource);
+                    num_resources += 1;
+                }
+                _ => {}
+            }
+        }
+        return Ok(num_resources);
+    }
+}
+
+
+impl ResourceError
+{
+    fn new(desc: String) -> Self
+    {
+        return ResourceError {
+            desc
+        }
+    }
+}
+
+
+impl Error for ResourceError
+{
+    fn description(&self) -> &str
+    {
+        return self.desc.as_str();
+    }
+}
+
+
+impl fmt::Display for ResourceError
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result
+    {
+        write!(f, "ResourceError({:?})", self.desc)
+    }
+}
+
+
+// Gets the extension of the path in uppercase.
+fn uppercase_extension(path: &Path) -> Result<String,Box<Error>>
+{
+    if let Some(extension) = path.extension() {
+        return Ok(extension.to_str().unwrap().to_uppercase());
```

Comment:
You could use `to_string_lossy` to get rid of the unwrap...

Reply:
This function is only called where the whole filename is known to have valid utf8, so the unwrap is avoiding the allocation of a temporary string. I'll keep the unwrap.

Also, there is no path that produces an error so I can return a String instead of a Result.